_source/_posts/2022-02-25-thymeleaf-security.md

```diff
@@ -0,0 +1,670 @@
+
+
+**This tutorial was created with the following frameworks and tools**:
+
+- [HTTPie 3.0.2](https://httpie.io/)
+- [Java OpenJDK 11](https://jdk.java.net/java-se-ri/11)
+- [Okta CLI 0.9.0](https://cli.okta.com)
+
+**Table of Contents**{: .hide }
+* Table of Contents
+{:toc}
+
+# What is Thymeleaf?
+
+Thymeleaf is an open-source server-side Java template engine for standalone and web applications, that was created by Daniel Fernández. The templates look like HTML and can be integrated with Spring MVC and Spring Security among other popular development frameworks. Although documentation about integration with Spring WebFlux is hard to find, it is currently supported and the Thymeleaf starter dependency performs the [auto-configuration](https://docs.spring.io/spring-framework/docs/current/reference/html/web-reactive.html#webflux-view-thymeleaf) of the template engine, template resolver and reactive view resolver.
+
+Among its features, Thymeleaf allows to:
+
+- Work with fragments: only a part of a template can be rendered, which can be useful for updating part of a page from responses in AJAX calls. It also provides a componentization tool, as the fragments can be included in multiple templates.
+- Handle forms with model objects containing their fields
+- Render variables and externalized text messages through its [Standard Expression Syntax](https://www.thymeleaf.org/doc/tutorials/3.0/usingthymeleaf.html#standard-expression-syntax)
+- Perform iterations and conditional evaluations
+
+{% img blog/thymeleaf-security/thymeleaf-logo-white.png alt:"Thymeleaf Logo" width:"600" %}{: .center-image }
+
+# Create a Spring Boot Webflux application with Thymeleaf
+
+Create a simple monolithic Spring Boot reactive application, that will use Thymeleaf for page templates.
+You can use [Spring Initializr](https://start.spring.io/), from its web UI, or download the starter app with the following HTTPie line:
+
+```shell
+https -d start.spring.io/starter.zip bootVersion==2.6.4 \
+  baseDir==thymeleaf-security \
+  groupId==com.okta.developer.thymeleaf-security \
+  artifactId==thymeleaf-security \
+  name==thymeleaf-security \
+  packageName==com.okta.developer.demo \
+  javaVersion==11 \
+  dependencies==webflux,okta,thymeleaf
+```
+
+Extract the Maven project, and some additional dependencies. `thymeleaf-extras-springsecurity5` dependency is required to include security conditionals in the templates, and
+`spring-security-test` has mock login features that will help when writing controller tests.
+
+```xml
+<dependency>
+  <groupId>org.thymeleaf.extras</groupId>
+  <artifactId>thymeleaf-extras-springsecurity5</artifactId>
+  <version>3.0.4.RELEASE</version>
+</dependency>
+<dependency>
+  <groupId>org.springframework.security</groupId>
+  <artifactId>spring-security-test</artifactId>
+  <scope>test</scope>
+</dependency>
```

Comment:
4 spaces for XML but it probably doesn't matter as long as you're consistent.

Reply:
Updated